subprojects/build-cache-http/src/integTest/groovy/org/gradle/caching/http/internal/NextGenHttpBuildCacheServiceTest.groovy

```diff
@@ -0,0 +1,143 @@
+import org.gradle.test.fixtures.server.http.HttpServer
+import org.gradle.util.TestUtil
+import org.junit.Rule
+import spock.lang.Specification
+
+class NextGenHttpBuildCacheServiceTest extends Specification {
+
+    @Rule
+    HttpServer server = new HttpServer()
+    @Rule
+    TestNameTestDirectoryProvider tempDir = new TestNameTestDirectoryProvider(getClass())
+
+    BuildCacheServiceFactory.Describer buildCacheDescriber
+    HttpClientHelper.Factory httpClientHelperFactory = HttpClientHelper.Factory.createFactory(new DocumentationRegistry())
+
+    def key = new DefaultBuildCacheKey(HashCode.fromString("01234567abcdef"))
+    private config = TestUtil.newInstance(HttpBuildCache.class)
+
+    NextGenHttpBuildCacheService cacheRef
+
+    NextGenHttpBuildCacheService getCache() {
+        if (cacheRef == null) {
+            buildCacheDescriber = new NoopBuildCacheDescriber()
+            cacheRef = new NextGenHttpBuildCacheServiceFactory(new DefaultSslContextFactory(), { it.addHeader("X-Gradle-Version", "3.0") }, httpClientHelperFactory)
+                .createBuildCacheService(this.config, buildCacheDescriber) as NextGenHttpBuildCacheService
+        }
+        cacheRef
+    }
+
+    def setup() {
+        server.start()
+        config.url = server.uri.resolve("/cache/")
+    }
+
+    def "stores cache artifact with nextGen key prefix"() {
+        def destFile = tempDir.file("cached.zip")
+        def content = "Data".bytes
+        server.expectPut("/cache/nextGen-${key.hashCode}", destFile, HttpStatus.SC_OK, null, content.length)
+
+        when:
+        cache.store(key, writer(content))
+
+        then:
+        destFile.bytes == content
+    }
+
+    def "loads artifact from cache with nextGen key prefix"() {
+        def srcFile = tempDir.file("cached.zip")
+        srcFile.text = "Data"
+        server.expectGet("/cache/nextGen-${key.hashCode}", srcFile)
+
+        when:
+        def receivedInput = null
+        cache.load(key) { input ->
+            receivedInput = input.text
+        }
+
+        then:
+        receivedInput == "Data"
+    }
+
+    def "checks if artifact exists in cache with nextGen key prefix"() {
+        def destFile = tempDir.file("cached.zip")
+        def content = "Data".bytes
+        server.expectPut("/cache/nextGen-${key.hashCode}", destFile, HttpStatus.SC_OK, null, content.length)
+        server.expectHead("/cache/nextGen-${key.hashCode}", destFile)
+        cache.store(key, writer(content))
+
+        when:
+        boolean exists = cache.contains(key)
+
+        then:
+        exists
+    }
+
+    private class NoopBuildCacheDescriber implements BuildCacheServiceFactory.Describer {
+
+        @Override
+        BuildCacheServiceFactory.Describer type(String type) { this }
+
+        @Override
+        BuildCacheServiceFactory.Describer config(String name, String value) { this }
+
+    }
+
+    static class Writer implements BuildCacheEntryWriter {
```

Comment:
🤔 Maybe it's time to create a `TestBuildCacheEntryWriter` and use that in our tests.